___

<a href='http://www.pieriandata.com'> <img src='../../Pierian_Data_Logo.png' /></a>
___

# SF Salaries Exercise 

Welcome to a quick exercise for you to practice your pandas skills! We will be using the [SF Salaries Dataset](https://www.kaggle.com/kaggle/sf-salaries) from Kaggle! Just follow along and complete the tasks outlined in bold below. The tasks will get harder and harder as you go along.

** Import pandas as pd.**

In [1]:
import pandas as pd

** Read Salaries.csv as a dataframe called sal.**

In [2]:
sal = pd.read_csv('Salaries.csv')

** Check the head of the DataFrame. **

In [3]:
sal.head(10)

Id        EmployeeName                                        JobTitle  \
0   1      NATHANIEL FORD  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY   
1   2        GARY JIMENEZ                 CAPTAIN III (POLICE DEPARTMENT)   
2   3      ALBERT PARDINI                 CAPTAIN III (POLICE DEPARTMENT)   
3   4   CHRISTOPHER CHONG            WIRE ROPE CABLE MAINTENANCE MECHANIC   
4   5     PATRICK GARDNER    DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)   
5   6      DAVID SULLIVAN                       ASSISTANT DEPUTY CHIEF II   
6   7           ALSON LEE              BATTALION CHIEF, (FIRE DEPARTMENT)   
7   8       DAVID KUSHNER                  DEPUTY DIRECTOR OF INVESTMENTS   
8   9      MICHAEL MORRIS              BATTALION CHIEF, (FIRE DEPARTMENT)   
9  10  JOANNE HAYES-WHITE          CHIEF OF DEPARTMENT, (FIRE DEPARTMENT)   

     BasePay  OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  \
0  167411.18         0.00  400184.25       NaN  567595.43         567595.43   
1  155966.02    245131.88  137811.38       NaN  538909.28         538909.28   
2  212739.13    106088.18   16452.60       NaN  335279.91         335279.91   
3   77916.00     56120.71  198306.90       NaN  332343.61         332343.61   
4  134401.60      9737.00  182234.59       NaN  326373.19         326373.19   
5  118602.00      8601.00  189082.74       NaN  316285.74         316285.74   
6   92492.01     89062.90  134426.14       NaN  315981.05         315981.05   
7  256576.96         0.00   51322.50       NaN  307899.46         307899.46   
8  176932.64     86362.68   40132.23       NaN  303427.55         303427.55   
9  285262.00         0.00   17115.73       NaN  302377.73         302377.73   

   Year  Notes         Agency  Status  
0  2011    NaN  San Francisco     NaN  
1  2011    NaN  San Francisco     NaN  
2  2011    NaN  San Francisco     NaN  
3  2011    NaN  San Francisco     NaN  
4  2011    NaN  San Francisco     NaN  
5  2011    NaN  San Francisco     NaN  
6  2011    NaN  San Francisco     NaN  
7  2011    NaN  San Francisco     NaN  
8  2011    NaN  San Francisco     NaN  
9  2011    NaN  San Francisco     NaN

<br></br>
These are the first ten entries.
<br></br>
<br></br>

** Use the .info() method to find out how many entries there are.**

In [4]:
sal.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 148654 entries, 0 to 148653
Data columns (total 13 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Id                148654 non-null  int64  
 1   EmployeeName      148654 non-null  object 
 2   JobTitle          148654 non-null  object 
 3   BasePay           148045 non-null  float64
 4   OvertimePay       148650 non-null  float64
 5   OtherPay          148650 non-null  float64
 6   Benefits          112491 non-null  float64
 7   TotalPay          148654 non-null  float64
 8   TotalPayBenefits  148654 non-null  float64
 9   Year              148654 non-null  int64  
 10  Notes             0 non-null       float64
 11  Agency            148654 non-null  object 
 12  Status            0 non-null       float64
dtypes: float64(8), int64(2), object(3)
memory usage: 14.7+ MB


<br></br>
There are 148654 entries.
<br></br>
<br></br>

**What is the average BasePay ?**

In [5]:
sal["BasePay"].mean()

66325.44884050643

<br></br>
The average BasePay is $66325.44884050643.
<br></br>
<br></br>

** What is the highest amount of OvertimePay in the dataset ? **

In [6]:
sal["OvertimePay"].max()

245131.88

<br></br>
The highest OvertimePay is $245131.88.
<br></br>
<br></br>

** What is the job title of  JOSEPH DRISCOLL ? Note: Use all caps, otherwise you may get an answer that doesn't match up (there is also a lowercase Joseph Driscoll). **

In [7]:
sal[sal["EmployeeName"] == "JOSEPH DRISCOLL"]

Id     EmployeeName                   JobTitle    BasePay  OvertimePay  \
24  25  JOSEPH DRISCOLL  CAPTAIN, FIRE SUPPRESSION  140546.86     97868.77   

    OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  Notes  \
24  31909.28       NaN  270324.91         270324.91  2011    NaN   

           Agency  Status  
24  San Francisco     NaN

In [8]:
sal[sal["EmployeeName"] == "JOSEPH DRISCOLL"]["JobTitle"]

24    CAPTAIN, FIRE SUPPRESSION
Name: JobTitle, dtype: object

<br></br>
JOSEPH DRISCOLL's JobTitle is CAPTAIN, FIRE SUPRESSION.
<br></br>
<br></br>

** How much does JOSEPH DRISCOLL make (including benefits)? **

In [9]:
sal[sal["EmployeeName"] == "JOSEPH DRISCOLL"]["TotalPayBenefits"]

24    270324.91
Name: TotalPayBenefits, dtype: float64

<br></br>
JOSEPH DRISCOLL's TotalPayBenefits are $270324.91
<br></br>
<br></br>

** What is the name of highest paid person (including benefits)?**

In [10]:
sal["TotalPayBenefits"].max()

567595.43

In [11]:
sal["TotalPayBenefits"] == sal["TotalPayBenefits"].max()

0          True
1         False
2         False
3         False
4         False
          ...  
148649    False
148650    False
148651    False
148652    False
148653    False
Name: TotalPayBenefits, Length: 148654, dtype: bool

In [12]:
sal[sal["TotalPayBenefits"] == sal["TotalPayBenefits"].max()]

Id    EmployeeName                                        JobTitle  \
0   1  NATHANIEL FORD  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY   

     BasePay  OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  \
0  167411.18          0.0  400184.25       NaN  567595.43         567595.43   

   Year  Notes         Agency  Status  
0  2011    NaN  San Francisco     NaN

<br></br>
NATHANIEL FORD has the highest TotalPayBenefits.
<br></br>
<br></br>

** What is the name of lowest paid person (including benefits)? Do you notice something strange about how much he or she is paid?**

In [13]:
sal["TotalPayBenefits"].min()

-618.13

In [14]:
sal["TotalPayBenefits"] == sal["TotalPayBenefits"].min()

0         False
1         False
2         False
3         False
4         False
          ...  
148649    False
148650    False
148651    False
148652    False
148653     True
Name: TotalPayBenefits, Length: 148654, dtype: bool

In [15]:
sal[sal["TotalPayBenefits"] == sal["TotalPayBenefits"].min()]

Id EmployeeName                    JobTitle  BasePay  OvertimePay  \
148653  148654    Joe Lopez  Counselor, Log Cabin Ranch      0.0          0.0   

        OtherPay  Benefits  TotalPay  TotalPayBenefits  Year  Notes  \
148653   -618.13       0.0   -618.13           -618.13  2014    NaN   

               Agency  Status  
148653  San Francisco     NaN

<br></br>
Joe Lopez has the lowest TotalPayBenefits.  His TotalPayBenefits seem odd because it is a negative number.
<br></br>
<br></br>

** What was the average (mean) BasePay of all employees per year? (2011-2014) ? **

In [16]:
year2011 = sal[sal["Year"] == 2011]["BasePay"].mean()

In [17]:
year2012 = sal[sal["Year"] == 2012]["BasePay"].mean()

In [18]:
year2013 = sal[sal["Year"] == 2013]["BasePay"].mean()

In [19]:
year2014 = sal[sal["Year"] == 2014]["BasePay"].mean()

In [20]:
print(f"Year 2011: {year2011}\nYear 2012: {year2012}\nYear 2013: {year2013}\nYear 2014: {year2014}")

Year 2011: 63595.956516774524
Year 2012: 65436.40685742255
Year 2013: 69630.0302164812
Year 2014: 66564.42192449933


In [21]:
sal.groupby("Year").mean()["BasePay"]

Year
2011    63595.956517
2012    65436.406857
2013    69630.030216
2014    66564.421924
Name: BasePay, dtype: float64

<br></br>
The mean BasePay in 2011 is $63595.956517.

The mean BasePay in 2012 is $65436.406857.

The mean BasePay in 2013 is $69630.030216.

The mean BasePay in 2014 is $66564.421924.
<br></br>
<br></br>

In [23]:
sal.groupby("JobTitle").mean()["BasePay"]

JobTitle
ACCOUNT CLERK                                     43300.806506
ACCOUNTANT                                        46643.172000
ACCOUNTANT INTERN                                 28732.663958
ACPO,JuvP, Juv Prob (SFERS)                       62290.780000
ACUPUNCTURIST                                     66374.400000
                                                      ...     
X-RAY LABORATORY AIDE                             47664.773077
X-Ray Laboratory Aide                             46086.387100
YOUTH COMMISSION ADVISOR, BOARD OF SUPERVISORS    52609.910000
Youth Comm Advisor                                39077.957500
ZOO CURATOR                                       43148.000000
Name: BasePay, Length: 2159, dtype: float64

In [24]:
sal.groupby("JobTitle").max()["BasePay"]

JobTitle
ACCOUNT CLERK                                     60838.20
ACCOUNTANT                                        65392.01
ACCOUNTANT INTERN                                 58501.72
ACPO,JuvP, Juv Prob (SFERS)                       62290.78
ACUPUNCTURIST                                     66374.40
                                                    ...   
X-RAY LABORATORY AIDE                             60086.62
X-Ray Laboratory Aide                             66123.01
YOUTH COMMISSION ADVISOR, BOARD OF SUPERVISORS    52609.91
Youth Comm Advisor                                57544.73
ZOO CURATOR                                       43148.00
Name: BasePay, Length: 2159, dtype: float64

In [25]:
sal.groupby("JobTitle").min()["BasePay"]

JobTitle
ACCOUNT CLERK                                       614.00
ACCOUNTANT                                            0.00
ACCOUNTANT INTERN                                  2981.53
ACPO,JuvP, Juv Prob (SFERS)                       62290.78
ACUPUNCTURIST                                     66374.40
                                                    ...   
X-RAY LABORATORY AIDE                              4690.68
X-Ray Laboratory Aide                              1007.00
YOUTH COMMISSION ADVISOR, BOARD OF SUPERVISORS    52609.91
Youth Comm Advisor                                13915.80
ZOO CURATOR                                       43148.00
Name: BasePay, Length: 2159, dtype: float64

<br></br>
There is a wide range in BasePay for JobTitle, but the mean BasePay seems similar for several JobTitle.
<br></br>
<br></br>

** How many unique job titles are there? **

In [26]:
sal.nunique()

Id                  148654
EmployeeName        110811
JobTitle              2159
BasePay             109489
OvertimePay          65998
OtherPay             83225
Benefits             98465
TotalPay            138486
TotalPayBenefits    142098
Year                     4
Notes                    0
Agency                   1
Status                   0
dtype: int64

In [30]:
jobTitlesList = sal['JobTitle'].tolist()

In [31]:
jobTitlesSet = set(jobTitlesList)

In [32]:
setCount = 0

for jobTitle in jobTitlesSet:
    setCount += 1

setCount

2159

In [27]:
sal.nunique()["JobTitle"]

2159

<br></br>
There are 2159 unique JobTitle.
<br></br>
<br></br>

** What are the top 5 most common jobs? **

In [28]:
sal["JobTitle"].value_counts()

Transit Operator                                  7036
Special Nurse                                     4389
Registered Nurse                                  3736
Public Svc Aide-Public Works                      2518
Police Officer 3                                  2421
                                                  ... 
CHIEF HOUSING INSPECTOR                              1
TRAFFIC SIGNAL OPERATOR                              1
COURT COMPUTER FACILITIES COORDINATOR                1
AUTOMOTIVE BODY AND FENDER WORKER SUPERVISOR I       1
VICTIM & WITNESS TECHNICIAN                          1
Name: JobTitle, Length: 2159, dtype: int64

In [29]:
sal["JobTitle"].value_counts().head(5)

Transit Operator                7036
Special Nurse                   4389
Registered Nurse                3736
Public Svc Aide-Public Works    2518
Police Officer 3                2421
Name: JobTitle, dtype: int64

<br></br>
The most common JobTitle are Transit Operator, Special Nurse, Registered Nurse, Public Svc Aide-Public Works, and Police Officer 3.
<br></br>
<br></br>

** How many Job Titles were represented by only one person in 2013? (e.g. Job Titles with only one occurence in 2013?) **

In [38]:
jobCountList = sal[sal["Year"] == 2013]["JobTitle"].value_counts().tolist()

In [39]:
jobCount = 0

for job in jobCountList:
    if job == 1:
        jobCount += 1

jobCount

202

In [33]:
sal[sal["Year"] == 2013]["JobTitle"].value_counts() == 1

Transit Operator                  False
Special Nurse                     False
Registered Nurse                  False
Public Svc Aide-Public Works      False
Custodian                         False
                                  ...  
Sr General Utility Mechanic        True
Sprv Adult Prob Ofc (SFERS)        True
Mayoral Staff XII                  True
Pr Investigator, Tax Collector     True
Senior Assessment Clerk            True
Name: JobTitle, Length: 1051, dtype: bool

In [34]:
sum(sal[sal["Year"] == 2013]["JobTitle"].value_counts() == 1)

202

<br></br>
In Year 2013 there were 202 JobTitle that only had a single person.
<br></br>
<br></br>

** How many people have the word Chief in their job title? (This is pretty tricky) **

In [36]:
jobTitlesList = sal['JobTitle'].tolist()

In [37]:
chiefCountList = 0

for jobTitle in jobTitlesList:
    if "chief" in jobTitle.lower():
        chiefCountList += 1

chiefCountList

627

<br></br>
There are 627 JobTitle with the word "Chief" in them.
<br></br>
<br></br>

** Bonus: Is there a correlation between length of the Job Title string and Salary? **

In [40]:
sal['JobTitle'].apply(len)

0         46
1         31
2         31
3         36
4         44
          ..
148649     9
148650    12
148651    12
148652    12
148653    26
Name: JobTitle, Length: 148654, dtype: int64

In [41]:
sal['JobTitle'].apply(len)

0         46
1         31
2         31
3         36
4         44
          ..
148649     9
148650    12
148651    12
148652    12
148653    26
Name: JobTitle, Length: 148654, dtype: int64

In [42]:
sal['title_len'] = sal['JobTitle'].apply(len)

In [43]:
sal['title_len']

0         46
1         31
2         31
3         36
4         44
          ..
148649     9
148650    12
148651    12
148652    12
148653    26
Name: title_len, Length: 148654, dtype: int64

In [45]:
sal[['BasePay','TotalPayBenefits']].corr()

BasePay  TotalPayBenefits
BasePay           1.000000          0.946595
TotalPayBenefits  0.946595          1.000000

In [46]:
sal[['TotalPay','TotalPayBenefits']].corr()

TotalPay  TotalPayBenefits
TotalPay          1.000000          0.977313
TotalPayBenefits  0.977313          1.000000

In [44]:
sal[['title_len','TotalPayBenefits']].corr()

title_len  TotalPayBenefits
title_len          1.000000         -0.036878
TotalPayBenefits  -0.036878          1.000000

<br></br>
There is no apparent correlation between the length of JobTitle and TotalPayBenefits.  There is a clear correlation between BasePay and TotalPayBenefits.  People that typically recieve more money as their base income tend to have higher total pay amounts, meaning that additional pay benefits do not make much of a difference for total pay, on average.  It might also mean that those with higher base pay receive similarly higher amounts of benefits.
<br></br>
<br></br>

# Great Job!